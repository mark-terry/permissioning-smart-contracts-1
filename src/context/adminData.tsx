// Libs
import React, { createContext, useContext, useMemo, useState, useEffect } from "react";
import { drizzleReactHooks } from "drizzle-react";

type ContextType = {
    admins?: string[],
    setAdmins?: (admins: string[]) => void
}

const AdminDataContext = createContext<ContextType>({})

export const AdminDataProvider: React.FC = (props: React.Props<{}>) => {
    const [admins, setAdmins] = useState<string[]>([])
    const value = useMemo(() => ({ admins, setAdmins }), [
        admins,
        setAdmins
    ])
    return <AdminDataContext.Provider value={value} {...props} />
}

export const useAdminData = () => {
    const context = useContext(AdminDataContext)

    if (!context) {
        throw new Error("useAccountData must be used within a AccountDataProvider.")
    }

    const { admins, setAdmins } = context

    const { userAddress } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
        userAddress: drizzleState.accounts[0]
    }));

    const { useCacheCall } = drizzleReactHooks.useDrizzle();    
    const adminList: string[] = useCacheCall("Admin", "getAdmins");

    // const { getAdmins } = drizzle.contracts.Admin.methods;

    useEffect(() => {
        // getAdmins().call().then((adminLists: string[]) => {setAdmins!(adminLists);});
        setAdmins!(adminList || [])
        }, [adminList, setAdmins]
    )
    

    const dataReady = useMemo(
        () =>
            Array.isArray(admins),
            [admins]
    );

    const isAdmin = useMemo(
        () => (dataReady && admins ? admins.includes(userAddress) : false),
        [dataReady, admins, userAddress]
    );

    const formattedAdmins = useMemo(() => {
        return admins
            ? admins
                  .map(address => ({
                      address,
                      identifier: address,
                      status: "active"
                  }))
                  .reverse()
            : undefined;
    }, [admins]);    

    return {
        dataReady,
        userAddress,
        isAdmin,
        admins: formattedAdmins
    }
}