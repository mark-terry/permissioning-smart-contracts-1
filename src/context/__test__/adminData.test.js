// Libs
import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
// Components
import { AdminDataProvider } from "../adminData";

describe("<AdminDataProvider />", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <AdminDataProvider>
                <div className="test" />
            </AdminDataProvider>
        );
    });

    it("renders children when passed in", () => {
        expect(wrapper.contains(<div className="test" />)).toEqual(true);
    });

    it("matches snapshot", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
