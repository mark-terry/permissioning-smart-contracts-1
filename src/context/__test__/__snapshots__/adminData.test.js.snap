// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminDataProvider /> matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "admins": Array [],
      "setAdmins": [Function],
    }
  }
>
  <div
    className="test"
  />
</ContextProvider>
`;
